lib/mm_game.rb

```diff
@@ -0,0 +1,125 @@
+require_relative 'mm_state'
+require_relative 'mm_outputs'
+require_relative 'mm_choice'
+require_relative 'mm_input'
+require_relative 'mm_code'
+require_relative 'mm_commands'
+
+DEFAULT = {
+    state: State,
+    game_in: Input,
+    game_out: Outputs,
+    choice: Choice,
+    code: Code,
+    commands: Commands
+}
+
+class Game
+    def initialize(options = {})
+        options = DEFAULT.merge(options)
+        @status = options[:state].new
+        @in = options[:game_in]
+        @out = options[:game_out]
+        @choice = options[:choice]
+        @code = options[:code]
+        @cmds = options[:commands]
+        @input_code = nil
+        @stop_game = false
+        @game_complete = false
+        play
+    end
+
+    def play
+        until @stop_game do
+            @game_complete = false
+            gameplay_loop
+        end
+        exit
+    end
+
+    def gameplay_loop
+        until @game_complete do
+            welcome_user
+            display_game_status
+            set_input_code(input_loop)
+            update_game_state
+            check_code_match
+            check_remaining_guesses
+        end
+    end
+
+    def welcome_user
+        if @status.state[:first_launch]
+            @out.welcome
+        end
+    end
+
+    def input_loop
+        user_in = @in.new
+        until user_in.is_valid do
+            if user_in.cmd_run != nil
```

Comment:
user_in.cmd_run w/o anything would work if it never returns false either. Otherwise, `unless user_in.cmd_run.nil? or `if !user_in.cmd_run.nil?`